src/crypto/index.js

```diff
@@ -936,25 +940,47 @@ Crypto.prototype.onSyncCompleted = async function(syncData) {
 };
 
 /**
- * Ask the server which users have new devices since a given token,
- * and invalidate them
+ * Trigger the appropriate invalidations and removes for a given
+ * device list
  *
- * @param {String} oldSyncToken
- * @param {String} lastKnownSyncToken
- *
- * Returns a Promise which resolves once the query is complete. Rejects if the
- *   keyChange query fails.
+ * @param {Object} deviceLists device_lists field from /sync, or response from
+ * /keys/changes
  */
-Crypto.prototype._invalidateDeviceListsSince = async function(
-    oldSyncToken, lastKnownSyncToken,
-) {
-    const r = await this._baseApis.getKeyChanges(
-        oldSyncToken, lastKnownSyncToken,
-    );
+Crypto.prototype._evalDeviceListChanges = async function(deviceLists) {
+    if (deviceLists.changed && Array.isArray(deviceLists.changed)) {
+        deviceLists.changed.forEach((u) => {
+            this._deviceList.invalidateUserDeviceList(u);
+        });
+    }
 
-    console.log("got key changes since", oldSyncToken, ":", r);
+    if (deviceLists.left && Array.isArray(deviceLists.left)) {
+        // Check we really don't share any rooms with these users
+        // any more: the server isn't required to give us the
+        // exact correct set.
+        const e2eUserIds = new Set(this._getE2eUsers());
 
-    await this.handleDeviceListChanges(r);
+        deviceLists.left.forEach((u) => {
+            if (!e2eUserIds.has(u)) {
```

Comment:
this looks O(N^2) ish

Reply:
The set membership shouldn't be O(n) though?

Reply:
er good point